tests/test_driver.py

```diff
@@ -0,0 +1,366 @@
+import collections
+
+import envi.const
+from fixtures import *
+
+import viv_utils.emulator_drivers as vudrv
+
+
+class LoggingMonitor(vudrv.Monitor):
+    """log the emulated addresses"""
+
+    def prehook(self, emu, op, startpc):
+        print("emu: 0x%x %s" % (startpc, op))
+
+    def preblock(self, emu, blockstart):
+        print("emu: block: start: 0x%x" % (blockstart))
+
+    def postblock(self, emu, blockstart, blockend):
+        print("emu: block: 0x%x - 0x%x" % (blockstart, blockend))
+
+
+class CoverageMonitor(vudrv.Monitor):
+    """capture the emulated addresses"""
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.addresses = collections.Counter()
+
+    def prehook(self, emu, op, startpc):
+        self.addresses[startpc] += 1
+
+
+def test_driver_monitor(pma01):
+    emu = pma01.getEmulator()
+    drv = vudrv.DebuggerEmulatorDriver(emu)
+    cov = CoverageMonitor()
+    drv.add_monitor(cov)
+
+    # 10001010 B8 F8 11 00 00          mov     eax, 11F8h
+    # 10001015 E8 06 02 00 00          call    __alloca_probe
+
+    drv.setProgramCounter(0x10001010)
+    drv.stepi()
+    assert drv.getProgramCounter() == 0x10001015
+
+    assert 0x10001010 in cov.addresses
+    assert 0x10001015 not in cov.addresses
+
+
+def test_dbg_driver_stepi(pma01):
+    emu = pma01.getEmulator()
+    drv = vudrv.DebuggerEmulatorDriver(emu)
+
+    # .text:10001342 57                      push    edi
+    # .text:10001343 56                      push    esi             ; fdwReason
+    # .text:10001344 53                      push    ebx             ; hinstDLL
+    # .text:10001345 E8 C6 FC FF FF          call    DllMain (0x10001010)
+    # .text:1000134A 83 FE 01                cmp     esi, 1
+    drv.setProgramCounter(0x10001342)
+    drv.stepi()
+    drv.stepi()
+    drv.stepi()
+    drv.stepi()
+    assert drv.getProgramCounter() == 0x10001010
+
+
+def test_dbg_driver_stepo(pma01):
+    emu = pma01.getEmulator()
+    drv = vudrv.DebuggerEmulatorDriver(emu)
+
+    # .text:10001342 57                      push    edi
+    # .text:10001343 56                      push    esi             ; fdwReason
+    # .text:10001344 53                      push    ebx             ; hinstDLL
+    # .text:10001345 E8 C6 FC FF FF          call    DllMain (0x10001010)
+    # .text:1000134A 83 FE 01                cmp     esi, 1
+    drv.setProgramCounter(0x10001342)
+    drv.stepo()
+    drv.stepo()
+    drv.stepo()
+    drv.stepo()
+    assert drv.getProgramCounter() == 0x1000134A
+
+
+class CreateMutexAHook:
+    """capture the mutex names passed to CreateMutexA"""
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.mutexes = set()
+
+    def __call__(self, emu, api, argv):
+        _, _, cconv, name, _ = api
```

Comment:
(1) would be helpful to have some helper/magic for this common pattern and the one below, see (2)